### PR TITLE
Handle missing design-time type metadata

### PR DESCRIPTION
**Pull request by bajtos**
- `@loopback/metadata`: Improve the typings for `getDesignTypeForProperty` and `getDesignTypeForMethod` to account for the fact that no metadata may be available when the project is written in pure JavaScript or not enabling TypeScript compiler option `emitDecoratorMetadata`.

- Other packages: Improve the code to handle the case where design-time type metadata is not available.

This work was triggered by a user reporting crash with unhelpful error message when their project did not have `emitDecoratorMetadata` enabled. 

Example error message:

```
Cannot start the application. TypeError: paramTypes is not iterable
  at resolveControllerSpec (loopback-next/packages/openapi-v3/dist/controller-spec.js:204:25)
  at Object.getControllerSpec (loopback-next/packages/openapi-v3/dist/controller-spec.js:323:16)
  at RestServer._createHttpHandler (loopback-next/packages/rest/dist/rest.server.js:289:42)
  (...)
```

**BREAKING CHANGE**

If you are consuming `@loopback/metadata` directly, then you may need to update your code to handle the case when design-time type metadata is not available. (The compiler will tell you what places to fix.)

Regular LoopBack users should not be affected as long as they update all LB packages together.

## Related issues & PRs:

- Implementing ts-mocha is throwing type errors in Inject #2764
- feat(context): add a check to see if design:type is emitted by TypeScript #2765

## Checklist

- [x] DCO (Developer Certificate of Origin) [signed in all commits](https://loopback.io/doc/en/contrib/code-contrib.html)
- [x] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- Documentation in [/docs/site](../tree/master/docs/site) was updated
- Affected artifact templates in `packages/cli` were updated
- Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: packages/metadata/src/types.ts [hunk 1]
```diff
@@ -77,13 +77,16 @@ export interface MetadataMap<T> {
  */
 export interface DesignTimeMethodMetadata {
   /**
-   * Type of the method itself. It is `Function`
+   * Type of the method itself. It is `Function` for methods, `undefined` for the constructor.
    */
```
Comment (bajtos) [inline packages/metadata/src/types.ts:81]: See the unit tests for examples of `DesignTimeMethodMetadata` values.

https://github.com/strongloop/loopback-next/blob/2f90a9531384c3e99afc9618d609b85ea9791e64/packages/metadata/src/__tests__/unit/inspector.unit.ts#L645-L676

File: packages/metadata/src/types.ts [hunk 2]
```diff
@@ -77,15 +77,18 @@ export interface MetadataMap<T> {
...
+
   parameterTypes: Function[];
+
   /**
-   * Return type
+   * Return type, may be `undefined` (e.g. for constructors).
    */
-  returnType: Function;
+  returnType: Function | undefined;
```
Comment (bajtos) [inline packages/metadata/src/types.ts:93]: I pushed this change in last minute, it turns out it breaks the build. I'll investigate tomorrow.

**Comment by bajtos** [comment]
I was running TypeScript in a watch mode and it looks like it did not catch all errors. The CI build is going to fail, I'll investigate & fix build errors tomorrow.

**Comment by bajtos** [comment]
However, early feedback is more than welcome!

File: packages/core/src/service.ts [hunk 3]
```diff
@@ -81,14 +82,28 @@ export function service(
...
           );
         }
       }
+      if (serviceType === undefined) {
+        const targetName =
+          typeof injection.target === 'function'
+            ? injection.target.name
+            : typeof injection.target === 'object'
+            ? injection.target.constructor.name
+            : inspect(injection.target);
```
Comment (raymondfeng) [inline packages/core/src/service.ts:99]: Please use https://loopback.io/doc/en/lb4/apidocs.metadata.decoratorfactory.gettargetname.html.

File: packages/metadata/src/inspector.ts [hunk 4]
```diff
@@ -235,6 +244,32 @@ export class MetadataInspector {
...
+      returnType === undefined
+    ) {
+      if (debug.enabled) {
+        const targetName =
+          typeof target === 'function'
+            ? target.name
+            : typeof target === 'object'
+            ? target.constructor.name
+            : inspect(target);
+        const propName = methodName ? `.${methodName}` : ' constructor';
```
Comment (raymondfeng) [inline packages/metadata/src/inspector.ts:261]: Ditto.

File: packages/core/src/service.ts [hunk 5]
```diff
@@ -81,14 +82,27 @@ export function service(
...
+        const targetName = DecoratorFactory.getTargetName(
+          injection.target,
+          injection.member,
+          injection.methodDescriptorOrParameterIndex,
+        );
+        const msg =
+          `No design-time type metadata found while inspecting ${targetName}. ` +
+          'You can either use `@service(ServiceClass)` or ensure `emitDecoratorMetadata` is enabled in your TypeScript configuration. ' +
+          'Run `tsc --showConfig` to print the final TypeScript configuration of your project.';
+        throw new Error(msg);
```
Comment (bajtos) [inline packages/core/src/service.ts:103]: Do we want to test this scenario?

File: packages/metadata/src/inspector.ts [hunk 6]
```diff
@@ -235,6 +244,25 @@ export class MetadataInspector {
...
+      if (debug.enabled) {
+        const targetName = DecoratorFactory.getTargetName(target, methodName);
+        debug(
+          'No design-time type metadata found while inspecting %s. ' +
+            'Did you forget to enable TypeScript compiler option `emitDecoratorMetadata`?',
+          targetName,
+        );
+      }
+
+      return undefined;
```
Comment (bajtos) [inline packages/metadata/src/inspector.ts:263]: Do we want to test this scenario?

File: packages/repository/src/decorators/model.decorator.ts [hunk 7]
```diff
@@ -87,6 +87,14 @@ export function buildModelDefinition(
     const propertyDef = propertyMap[p];
     const designType = MetadataInspector.getDesignTypeForProperty(prototype, p);
     if (!propertyDef.type) {
+      if (!designType) {
+        throw new Error(
+          `The definition of model property ${modelDef.name}.${p} is missing ` +
+            '`type` field and TypeScript compiler option `emitDecoratorMetadata` ' +
+            'is not set. Please enable `emitDecoratorMetadata` or specify ' +
+            'the type explicitly.',
```
Comment (bajtos) [inline packages/repository/src/decorators/model.decorator.ts:95]: Do we want to test this scenario?

**Review by bajtos — COMMENTED** [review]
I fixed compiler errors and addressed review comments in e396d09.

It makes me wonder if I should try to write tests for the three new branches I introduced? I think it's possible to create a class with LB decorators but without TS design-time metadata, but I am not sure if it's worth the effort.

Something along the following lines - the trick is to prevent TypeScript from transforming our code building the class:

```ts
import {model, property} from '@loopback/repository';

function createModelWithNoDesignTimeMetadata() {
  const factory = new Function('model', 'property', `
    class MyModel {
    };
    
    property({id: true})(MyModel, 'id');
    model()(MyModel);
    
    return MyModel;
`);

  return factory(model, property);
}
```

File: packages/core/src/service.ts (hunk 5 above)
Comment (raymondfeng) [reply packages/core/src/service.ts:103]: As long as you have one test to cover the case where `design time` metadata is missing, it's good enough for me. Your code to simulate such cases looks reasonable.

**Comment by bajtos** [comment]
In 6dd386a, I am adding two unit tests, one for `getDesignTypeForProperty` and the other for `getDesignTypeForMethod`.

@raymondfeng LGTY now?

(I'll clean up the commit history before landing.)

**Comment by bajtos** [comment]
Related:
- Implementing ts-mocha is throwing type errors in Inject #2764
- feat(context): add a check to see if design:type is emitted by TypeScript #2765

**Review by raymondfeng — APPROVED** [review]

**Review by raymondfeng — CHANGES_REQUESTED** [review]
Please fix the CI test failure.

**Comment by bajtos** [comment]
> Please fix the CI test failure.

It turns out there are valid cases where `emitDecoratorMetadata` is enabled but design-time metadata is not available:
- Complex types like arrays (`string[]`) or unions (string | number)
- Circular references between classes (e.g. for `@belongsTo` relationship)

I pushed two new commits to address the newly discovered aspects.
1. `@loopback/metadata`: add tests for complex types, improve TSDoc comment
2. Improve handling of missing design-time type in `@loopback/repository` and `@loopback/repository-json-schema`. Add a comprehensive description of the new error to https://loopback.io/doc/en/lb4/Error-codes.html.

@raymondfeng PTAL again.

File: docs/site/reference/error-codes.md [hunk 8]
```diff
@@ -16,6 +16,7 @@ sets the error `code` property to a machine-readable string.
 | VALIDATION_FAILED          | The data provided by the client is not a valid entity.                                                                                                                                                    |
 | INVALID_PARAMETER_VALUE    | The value provided for a parameter of a REST endpoint is not valid. For example, a string value was provided for a numeric parameter.                                                                     |
 | MISSING_REQUIRED_PARAMETER | No value was provided for a required parameter.                                                                                                                                                           |
+| CANNOT_INFER_PROPERTY_TYPE | See below: [CANNOT_INFER_PROPERTY_TYPE](#cannot_infer_property_type)                                                                                                                                      |
```
Comment (bajtos) [inline docs/site/reference/error-codes.md:19]: Maybe we should get rid of this table and create a level2 (`##`) section for each error code? There will be auto-generated table of contents at the top of the page to allow quick lookup of a particular error.

I prefer to make such changes in a different PR, preferably after this one is landed.

File: docs/site/reference/error-codes.md [hunk 9]
```diff
@@ -26,3 +27,70 @@ See the following resources for a list of low-level error codes:
...
+- The property has a type not supported by TypeScript metadata engine:
+  - `undefined`
+  - `null`
+  - complex types like arrays (`string[]`), generic types (`Partial<MyModel>`),
+    union types (`string | number`), and more.
+- The TypeScript project has not enabled the compiler option
+  `emitDecoratorMetadata`.
+- The code is written in vanilla JavaScript.
+
+<a id="cannot_infer_property_type-solutions"></a>
```
Comment (bajtos) [inline docs/site/reference/error-codes.md:58]: This html snippet creates a unique anchor. I think `### Solutions` creates the same anchor `#solutions` for all error codes, which is not useful.

**Review by raymondfeng — APPROVED** [review]

File: docs/site/reference/error-codes.md (hunk 8 above)
Comment (bajtos) [reply docs/site/reference/error-codes.md:19]: Opened https://github.com/strongloop/loopback-next/pull/6363
